evm/v0.5/src/helpers.ts

```diff
@@ -603,24 +665,183 @@ export function generateSAID(sa: ServiceAgreement): Hash {
+  )
+
+  const sAID = generateSAID(serviceAgreement)
+  const signatures = []
+
+  for (let i = 0; i < signers.length; i++) {
+    const signer = signers[i]
+    const oracle = serviceAgreement.oracles[i]
+    const oracleSignature = await personalSign(sAID, signer)
+    const requestDigestAddr = recoverPersonalSignature(sAID, oracleSignature)
+    assert.equal(oracle, requestDigestAddr)
+    signatures.push(oracleSignature)
+  }
+
+  return combineOracleSignatures(signatures)
+}
+
+/**
+ * Combine v, r, and s params of multiple signatures into format expected by contracts
+ * @param signatures The list of signatures to combine
+ */
+export function combineOracleSignatures(
+  signatures: Signature[],
+): OracleSignatures {
+  return {
+    vs: signatures.map(os => os.v),
+    rs: signatures.map(os => os.r),
+    ss: signatures.map(os => os.s),
+  }
+}
+
+/**
+ * Signs a message
+ * @param message The message to sign signed
+ * @param wallet The wallet of the signer
+ */
+export async function personalSign(
+  message: any,
+  wallet: ethers.Wallet,
+): Promise<Signature> {
+  const eMsg =
+    `Message ${message} is not a recognized representation of a ` +
+    'byte array. (Can be Buffer, BigNumber, Uint8Array, 0x-prepended ' +
+    'hexadecimal string.)'
+  assert(isByteRepresentation(message), eMsg)
+  const flatSig = await wallet.signMessage(utils.arrayify(message))
+  const splitSignature = utils.splitSignature(flatSig)
+  const { v, r, s } = splitSignature
+  if (!v) {
+    throw Error(`Could not extract v from signature`)
+  } else {
+    return { v, r, s }
+  }
+}
+
+/**
+ * Recovers the address of the signer of a message
+ * @param message The message that was signed
+ * @param signature The signature on the message
+ */
+export function recoverPersonalSignature(
+  message: any,
+  signature: Signature,
+): any {
+  const digest = utils.arrayify(message)
+  return utils.verifyMessage(digest, signature)
+}
+
+/** Check thathe given service agreement was stored at the correct location */
+export const checkServiceAgreementPresent = async (
+  coordinator: Instance<CoordinatorFactory>,
+  serviceAgreement: ServiceAgreement,
+) => {
+  const sAID = generateSAID(serviceAgreement)
+  const sa = await coordinator.serviceAgreements(sAID)
+  assertBigNum(sa[0], bigNum(serviceAgreement.payment), 'expected payment')
+  assertBigNum(
+    sa[1],
+    bigNum(serviceAgreement.expiration),
+    'expected expiration',
+  )
+  assertBigNum(sa[2], bigNum(serviceAgreement.endAt), 'expected endAt date')
+  assert.equal(sa[3], serviceAgreement.requestDigest, 'expected requestDigest')
+}
+
+// Check that all values for the struct at this SAID have default values. I.e.
+// nothing was changed due to invalid request
+export const checkServiceAgreementAbsent = async (
+  coordinator: Instance<CoordinatorFactory>,
+  serviceAgreementID: string,
+) => {
+  const sa = await coordinator.serviceAgreements(
+    toHex(serviceAgreementID).slice(0, 66),
```

Comment:
Could we use a comment here saying that we're slicing because the serviceAggrementId is contained within the highest 32 bytes?